pkg/sentry/socket/hostinet/socket_unsafe.go

```
@@ -55,8 +56,116 @@ func writev(fd int, srcs []unix.Iovec) (uint64, error) {
+	for index, iface = range stk.Interfaces() {
+		if iface.Name == ifr.Name() {
+			found = true
+			break
+		}
+	}
+	if !found {
+		return linuxerr.ENODEV
+	}
+
+	switch arg {
+	case linux.SIOCGIFINDEX:
+		// Copy out the index to the data.
+		hostarch.ByteOrder.PutUint32(ifr.Data[:], uint32(index))
+
+	case linux.SIOCGIFADDR:
+		// Copy the IPv4 address out.
+		for _, addr := range stk.InterfaceAddrs()[index] {
+			// This ioctl is only compatible with AF_INET addresses.
+			if addr.Family != linux.AF_INET {
+				continue
+			}
+			copy(ifr.Data[4:8], addr.Addr)
+			break
+		}
+
+	case linux.SIOCGIFHWADDR:
+		// Copy the hardware address out.
+		//
+		// Refer: https://linux.die.net/man/7/netdevice
+		// SIOCGIFHWADDR, SIOCSIFHWADDR
+		//
+		// Get or set the hardware address of a device using
+		// ifr_hwaddr. The hardware address is specified in a struct
+		// sockaddr. sa_family contains the ARPHRD_* device type,
+		// sa_data the L2 hardware address starting from byte 0. Setting
+		// the hardware address is a privileged operation.
+		hostarch.ByteOrder.PutUint16(ifr.Data[:], iface.DeviceType)
+		n := copy(ifr.Data[2:], iface.Addr)
+		for i := 2 + n; i < len(ifr.Data); i++ {
+			ifr.Data[i] = 0 // Clear padding.
+		}
+
+	case linux.SIOCGIFMTU:
+		// Gets the MTU of the device.
+		hostarch.ByteOrder.PutUint32(ifr.Data[:4], iface.MTU)
+
+	case linux.SIOCGIFNETMASK:
+		// Gets the network mask of a device.
+		for _, addr := range stk.InterfaceAddrs()[index] {
+			// This ioctl is only compatible with AF_INET addresses.
+			if addr.Family != linux.AF_INET {
+				continue
+			}
+			// Populate ifr.ifr_netmask (type sockaddr).
+			hostarch.ByteOrder.PutUint16(ifr.Data[0:], uint16(linux.AF_INET))
+			hostarch.ByteOrder.PutUint16(ifr.Data[2:], 0)
+			var mask uint32 = 0xffffffff << (32 - addr.PrefixLen)
+			// Netmask is expected to be returned as a big endian
+			// value.
+			binary.BigEndian.PutUint32(ifr.Data[4:8], mask)
+			break
+		}
```

If the loop completes without finding an AF_INET address, we should probably return an error.

---

I returned linuxerr.EADDRNOTAVAIL both in socket_unsafe.go and netstack.go according to https://github.com/torvalds/linux/blob/v5.19/net/ipv4/devinet.c#L1111

---

As above, this will always return `EADDRNOTAVAIL` because the `break` only breaks out of the for loop.

---

Done, my negligence